vortex-array/src/data.rs

```diff
@@ -155,16 +155,17 @@ impl Statistics for ArrayData {
             return Some(s);
         }
 
+        let computed = self
+            .to_array()
+            .with_dyn(|a| a.compute_statistics(stat))
+            .ok()?;
+
         self.stats_map
             .write()
             .unwrap_or_else(|_| {
                 vortex_panic!("Failed to write to stats map while computing {}", stat)
             })
-            .extend(
-                self.to_array()
-                    .with_dyn(|a| a.compute_statistics(stat))
-                    .ok()?,
-            );
+            .extend(computed);
```

Comment:
doing it this way means that computing the stats happens when we aren't holding a write lock, which allows us to use the `self.into_canonical` trick in a couple of places to compute stats... otherwise we deadlock